### ISSUE TITLE
Symmetry Enhanced Variational Quantum Spin Eigensolver

### ISSUE THREAD

**Issue by NadavClassiq** [Classiq/classiq-library#577]
#### Abstract
Variational Quantum Algorithms (VQAs) have emerged as powerful methods for achieving quantum advantage on Noise Intermediate-Scale Quantum (NISQ) devices. Their adaptability to current quantum technology, without the need for error correction, makes them ideal for a range of applications. A key component in building effective VQAs is tailoring the variational ansatz to include symmetries or conserved quantities, which can improve performance and accuracy. This project focuses on implementing the [Symmetry Enhanced Variational Quantum Spin Eigensolver](https://quantum-journal.org/papers/q-2023-01-19-899/) by Chufan Lyu et al., specifically designed for spin systems with a conserved total spin number using the Variational Quantum Eigensolver (VQE).

### Project Overview

**Challenge:** Implement the VQE algorithm using a symmetry-enhanced ansatz on the Classiq platform to estimate ground state energies for a spin system. Compare the **Sz-conserving ansatz** with **hardware-efficient ansatzes** between 5 and 15 layers. Reference the [Glued Trees example structure](https://github.com/Classiq/classiq-library/blob/main/research/glued_trees/glued_trees.ipynb) for notebook organization.

### Objective

**Estimate the ground state energy for a spin system with  $N = 4$ qubits**:

1. Implement and compare:
   - Sz-conserving ansatz
   - Hardware-efficient ansatzes (5 to 15 layers)
2. Analyze the algorithm’s performance with different initial parameters.

### Deliverables

- **Jupyter Notebook** containing:
   - Quantum programs implementing the VQE with both ansatz types.
   - Graphical comparison of the ground state energy estimates for each ansatz.
   - Analysis of algorithm behavior with varying initial parameters.

Follow the **Contribution Guidelines** in [CONTRIBUTING.md](https://github.com/Classiq/classiq-library/blob/main/CONTRIBUTING.md). Reach out on GitHub or in our [Slack Community](https://classiq-community.slack.com/join/shared_invite/zt-2s4gsf6gi-m0YIoI80fyzTaIeVBbmSOw#/shared-invite/email) if you have any questions.

### Getting Started

1. **Review Paper:** Study the theory and methods from the paper [Symmetry Enhanced Variational Quantum Spin Eigensolver](https://quantum-journal.org/papers/q-2023-01-19-899/) by Chufan Lyu et al.
2. **Set Up Environment:** Create a new Jupyter Notebook and install the Classiq SDK. Follow the [setup guide](https://docs.classiq.io/latest/classiq_101/registration_installations/#python-sdk-installation).
3. **Guiding Materials:**
   - [Classiq 101](https://docs.classiq.io/latest/classiq_101/) - Introduction to platform concepts.
   - [Classiq Fundamentals Workshop](https://www.youtube.com/watch?v=k9Lqz6r-8A4&list=PL_wGNAk5B0pXq98BJBKsbNH2Qjp1lk8dG&index=16) - Hands-on Classiq basics.
   - Platform walkthrough on [home page](https://platform.classiq.io/).

### Implementation Steps

1. **Algorithm Coding:**  
   - Implement the algorithm using the Classiq SDK.
   - Document steps in markdown, following the [Glued Trees Example](https://github.com/Classiq/classiq-library/blob/main/research/glued_trees/glued_trees.ipynb).
   - For assistance, reach out on GitHub or [Slack](https://short.classiq.io/join-slack).

2. **Mathematical Explanation:**  
   - Use markdown and LaTeX for theoretical background, key equations, and algorithm insights.

3. **Generate `.qmod` File:**  
   - Use `write_qmod(model, "filename.qmod")` in your code to save your models.
   - Confirm successful execution and `.qmod` file generation.

4. **Quality Check:**  
   - Proofread and ensure code accuracy.
   - Format markdown for clarity and professional presentation.

5. **Submit Contribution:**  
   - Follow [Contribution Guidelines](https://github.com/Classiq/classiq-library/blob/main/CONTRIBUTING.md).
   - Open a Pull Request in `classiq-library/research/symmetry_enhanced_variational_spin_eigensolver`.
   - Include a summary and key insights from your implementation.

---

### Resources

- **Reference Paper:** [Symmetry Enhanced Variational Quantum Spin Eigensolver by Chufan](https://quantum-journal.org/papers/q-2023-01-19-899/) Lyu et al.
- **Glued Trees Example:** [Link](https://github.com/Classiq/classiq-library/blob/main/research/glued_trees/glued_trees.ipynb)
- **Contribution Guidelines:** [CONTRIBUTING.md](https://github.com/Classiq/classiq-library/blob/main/CONTRIBUTING.md)

---

**Note:** No strict deadline. Confirm with us if you start this task so we can assign it to you.

**Good Luck!**

**Comment by KYC2654** [comment]
Hi, I'd like to be assigned to this project

**Comment by NadavClassiq** [comment]
Hi @KYC2654! 

Yes, I am assigning you.

Please note that we accept high-quality implementations to our repository and will be glad to accept a contribution that meets our standards.

Good luck!

**Comment by TaliCohn** [comment]
Hi @KYC2654, what is the status of this? Are you still working on the implementation?

**Comment by KYC2654** [comment]
Yes! Deeply apologize for the delay, but I am still working on it. I'm having a tough time trying to implement the ansatzs. I realized the deadline was last week, but I need more time.

**Comment by TaliCohn** [comment]
Sure, how much time do you still need? 
For ant questions, please feel free to post on the Community Slack - out=r team and community can help you out.

**Comment by KYC2654** [comment]
Perhaps an additional week or two. Sorry for the late reply and delay! Still in school, and it's midterm season.

**Comment by TaliCohn** [comment]
Hi @KYC2654, what is the status here? Are you still working on this? 
If you have opened a PR, please link it to this issue so we can keep track.
